router/handler_edge_ctrl/apiSessionAdded.go

```diff
@@ -43,15 +64,233 @@ func (h *apiSessionAddedHandler) HandleReceive(msg *channel2.Message, ch channel
+	}
+
+	//if no id or the sync id is newer, reset
+	if h.syncTracker == nil || h.syncTracker.syncId == "" || h.syncTracker.syncId < state.Id {
+		logger.Warnf("new syncId [%s], resetting", state.Id)
+
+		if h.syncTracker != nil {
+			h.syncTracker.Stop()
+		}
+
+		h.syncTracker = newApiSessionSyncTracker(state.Id)
+
+		h.syncTracker.StartDeadline(h.syncReady, h.syncFail, 20*time.Second)
+	}
+
+	//ignore older syncs
+	if h.syncTracker.syncId > state.Id {
+		logger.Warnf("older syncId [%s], ignoring", state.Id)
+		return
+	}
+
+	h.syncTracker.Add(reqWithState)
+
+}
+
+type apiSessionSyncTracker struct {
+	syncId           string
+	syncLastRecieved bool
+	reqsWithState    map[int]*apiSessionAddedWithState
+	hasLast          bool
+	lastSeq          int
+	stop             chan struct{}
+	deadline         sync.Once
+	isDone           bool
+}
+
+func newApiSessionSyncTracker(id string) *apiSessionSyncTracker {
+	return &apiSessionSyncTracker{
+		syncId:        id,
+		reqsWithState: map[int]*apiSessionAddedWithState{},
+		stop:          make(chan struct{}, 0),
+	}
+}
+
+func (tracker *apiSessionSyncTracker) Add(reqWithState *apiSessionAddedWithState) {
+	tracker.reqsWithState[reqWithState.Sequence] = reqWithState
+
+	if reqWithState.IsLast {
+		tracker.hasLast = true
+		tracker.lastSeq = reqWithState.Sequence
+	}
+}
+
+func (tracker *apiSessionSyncTracker) Stop() {
+	if tracker != nil {
+		tracker.stop <- struct{}{}
+	}
+}
+
+func (tracker *apiSessionSyncTracker) StartDeadline(syncReady chan []*edge_ctrl_pb.ApiSession, syncFail chan error, timeout time.Duration) {
+	tracker.deadline.Do(func() {
+		go func() {
+			ticker := time.NewTicker(1 * time.Second)
+			select {
+			case <-tracker.stop:
+				tracker.reqsWithState = nil
+				syncFail <- nil
+				return
+			case <-ticker.C:
+				if tracker.HasAll() {
+					syncReady <- tracker.all()
+					return
+				}
+			case <-time.After(timeout):
+				tracker.reqsWithState = nil
+				syncFail <- errors.New("timeout, did not receive all updates in time")
+				return
+			}
+		}()
+	})
+}
+
+func (tracker *apiSessionSyncTracker) HasAll() bool {
+	if !tracker.hasLast {
+		return false
+	}
+
+	hasAll := true
+
+	for i := 0; i <= tracker.lastSeq; i++ {
+		if req, ok := tracker.reqsWithState[i]; !ok && req == nil {
+			hasAll = false
+			break
+		}
+	}
+
+	return hasAll
+}
+
+func (tracker *apiSessionSyncTracker) all() []*edge_ctrl_pb.ApiSession {
+	var result []*edge_ctrl_pb.ApiSession
+	for i := 0; i <= tracker.lastSeq; i++ {
+		if req, ok := tracker.reqsWithState[i]; ok {
+			for _, apiSession := range req.ApiSessions {
+				result = append(result, apiSession)
+			}
+		} else {
+			pfxlog.Logger().WithField("strategy", sync_strats.RouterSyncStrategyInstant).Error("all failed to have all update sequences")
+		}
+	}
+
+	return result
+}
+
+type apiSessionAddedWithState struct {
+	SyncStrategyType string
+	*sync_strats.InstantSyncState
+	*edge_ctrl_pb.ApiSessionAdded
+}
+
```

Comment:
type is unused

Reply:
I don't understand. I looked up usages and it is used.

Reply:
Oh the type below. Removed.